### PR TITLE
Feature/brainalive

### PR DESCRIPTION
**Pull request by dioptre**
Moving https://github.com/neuromore/studio/pull/105 here.

Pinning to commit dc9149b

Everything seems functional. Awaiting final code review and push @cyberjunk 

Tested on Mac, Windows

### CODE REVIEW

**Comment by cyberjunk** [comment]
@dioptre @Sarthak031
I started testing all builds.
Unfortunately, I do not have the device to run a full test...

**1) Make on OSX/Ubuntu 18.04**
Worked out of box (this is equal to `Release+X64` in VS)

**2) VS 2019 on Windows 10**
All configurations Worked out of box except for configuration `Production-ANT` which is the only one with `INCLUDE_DEVICE_BRAINALIVE` not defined (support for BrainAlive disabled). I needed the two fixes above to make it build:

https://github.com/neuromore/studio/pull/113/commits/4e1dcdeba14f4d2f703b6e33a8a3edd72cbb1c25 is trivial, if the device is disabled, the .moc files should not be built, so I excluded them like done for other disabled devices.
https://github.com/neuromore/studio/pull/113/commits/22346bf3fa2f925853246daeda6cfcc470227e22 is more interesting. It seems `mData_2` is used in [GraphAttributesWidget.cpp](https://github.com/neuromore/studio/blob/22346bf3fa2f925853246daeda6cfcc470227e22/src/Studio/Plugins/Graph/GraphAttributesWidget.cpp#L35) which is a bit confusing on first glance. Because GraphAttributesWidget seems generic and not related to a specific device but DOES require this global c var from BrainAlive Device Code (which is very specific and built optionaly). I might request some changes regarding this (e.g. moving mData_2 out of BrainAlive context/folder if it is something that GraphAttributesWidget depends on), but I need to do a full code review first to get some better idea.

**Comment by dioptre** [comment]
Great thanks mate.

**Comment by Sarthak031** [comment]
> @dioptre @Sarthak031
> I started testing all builds.
> Unfortunately, I do not have the device to run a full test...
> 
> **1) Make on OSX/Ubuntu 18.04**
> Worked out of box (this is equal to `Release+X64` in VS)
> 
> **2) VS 2019 on Windows 10**
> All configurations Worked out of box except for configuration `Production-ANT` which is the only one with `INCLUDE_DEVICE_BRAINALIVE` not defined (support for BrainAlive disabled). I needed the two fixes above to make it build:
> 
> [4e1dcde](https://github.com/neuromore/studio/commit/4e1dcdeba14f4d2f703b6e33a8a3edd72cbb1c25) is trivial, if the device is disabled, the .moc files should not be built, so I excluded them like done for other disabled devices.
> [22346bf](https://github.com/neuromore/studio/commit/22346bf3fa2f925853246daeda6cfcc470227e22) is more interesting. It seems `mData_2` is used in [GraphAttributesWidget.cpp](https://github.com/neuromore/studio/blob/22346bf3fa2f925853246daeda6cfcc470227e22/src/Studio/Plugins/Graph/GraphAttributesWidget.cpp#L35) which is a bit confusing on first glance. Because GraphAttributesWidget seems generic and not related to a specific device but DOES require this global c var from BrainAlive Device Code (which is very specific and built optionaly). I might request some changes regarding this (e.g. moving mData_2 out of BrainAlive context/folder if it is something that GraphAttributesWidget depends on), but I need to do a full code review first to get some better idea.

@cyberjunk I have updated your suggested changes but the problem is where I can commit my changes because @dioptre  closed my pull request. And the "[Feature/brainalive](https://github.com/neuromore/studio/pull/113)" has not to write access for me.
Am I need to create a pull request again?

**Comment by dioptre** [comment]
Hi @Sarthak031 I already mentioned that you need to do another PR in an email.

Either that or show me your recent commits.

Thanks
Andrew

**Comment by Sarthak031** [comment]
> Hi @Sarthak031 I already mentioned that you need to do another PR in an email.
> 
> Either that or show me your recent commits.
> 
> Thanks
> Andrew

@dioptre @cyberjunk  I have created the new pull request according to your suggestions.

**Comment by dioptre** [comment]
Building for me on Mac @cyberjunk there's quite a lot changed in there, but seems good.

File: src/Studio/Plugins/Graph/GraphAttributesWidget.h [hunk 1]
```diff
@@ -114,17 +118,41 @@ class GraphAttributesWidget : public QScrollArea, public Core::EventHandler
...
+        void dataReceived(QByteArray data);
+        void On_connect();
+        void On_Ok();
 		uint32 FindRemoveButtonIndex(QObject* button) const;
 
 		GraphObject*					mGraphObject;
 		GraphObjectViewWidget*			mGraphObjectView;
 		PropertyTreeWidget*				mPropertyTreeWidget;
 		Property*						mNameProperty;
+		Property*						mNameProperty_1;
```
Comment (cyberjunk) [inline src/Studio/Plugins/Graph/GraphAttributesWidget.h:130]: Please init these new pointer fields to null in the constructor like done for `mNameProperty`.
Please also double check it in other classes.

File: src/Studio/Devices/BrainAlive/BrainAliveDriver.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	// for autodetection
+	QThread* mAutoDetectionThread_2;
+	QTimer* mAutoDetectionTimer_2;
+	
+
+	// list to keep track of connected devices (index-matched with theh serial thread array below)
+	Core::Array<BrainAliveDeviceBase*> mDevices;
+
+	// one thread per device
+	// TODO use one thread for all serial handlers
```
Comment (cyberjunk) [inline src/Studio/Devices/BrainAlive/BrainAliveDriver.h:89]: Please double check all your TODO remarks (there are quite a few..).
Which of them are still relevant and which of them are critical?

**Review by cyberjunk — CHANGES_REQUESTED** [review]
It builds fine on all platforms, but there are still a few things that worry me a bit. Additionally to the specific comments, it's about your usage of global-like c variables with the `extern` keyword and sharing them not only across classes but also across project/library boundaries. For instance, the already discussed `mData_2` (but there are more). This isn't something that has been done in the project this way so far. So my questions regarding this are mainly:

1) Why didn't you decide in favor of some more C++/OOP based approach, like static properties on classes or a classic singleton approach (we got this architecture pattern and many singleton instances already in the project)

2) Are you absolutely sure the data stored in these variables isn't something that should be rather exist per device (per node) instance and hence be a regular property on some class? Is or will there be support for having multiple BrainAlive devices connected at the same time? (This is something we DO support for some other devices, so it must not break in general)

**Pull request closed without merging**